users/src/main/java/io/crate/operation/user/UserManagerService.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * This file is part of a module with proprietary Enterprise Features.
+ *
+ * Licensed to Crate.io Inc. ("Crate.io") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ *
+ * Unauthorized copying of this file, via any medium is strictly prohibited.
+ *
+ * To use this file, Crate.io must have given you permission to enable and
+ * use such Enterprise Features and you must have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  If you enable or use the Enterprise
+ * Features, you represent and warrant that you have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  Your use of the Enterprise Features
+ * if governed by the terms and conditions of your Enterprise or Subscription
+ * Agreement with Crate.io.
+ */
+
+package io.crate.operation.user;
+
+import com.google.common.collect.ImmutableSet;
+import io.crate.action.FutureActionListener;
+import io.crate.action.sql.SessionContext;
+import io.crate.analyze.*;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.relations.AnalyzedRelationVisitor;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Nullable;
+
+import java.util.Locale;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import static io.crate.operation.user.UsersMetaData.PROTO;
+import static io.crate.operation.user.UsersMetaData.TYPE;
+
+public class UserManagerService implements UserManager, ClusterStateListener {
+
+    static User CRATE_USER = new User("crate", true);
+
+    static {
+        MetaData.registerPrototype(TYPE, PROTO);
+    }
+
+    private final TransportCreateUserAction transportCreateUserAction;
+    private final TransportDropUserAction transportDropUserAction;
+    private final PermissionVisitor permissionVisitor = new PermissionVisitor();
+    private volatile Set<User> users = ImmutableSet.of(CRATE_USER);
+
+    public UserManagerService(TransportCreateUserAction transportCreateUserAction,
+                              TransportDropUserAction transportDropUserAction,
+                              ClusterService clusterService) {
+        this.transportCreateUserAction = transportCreateUserAction;
+        this.transportDropUserAction = transportDropUserAction;
+        clusterService.add(this);
+    }
+
+    static Set<User> getUsersFromMetaData(@Nullable UsersMetaData metaData) {
```

Comment:
I'd rename this to just `getUsers` - the `fromMetaData` is strange given that the argument can be null and you actually get a user back if it's null.